src/py/flwr/client/process/clientappio_servicer.py

```diff
@@ -0,0 +1,147 @@
+        """Pull Message, Context, and Run."""
+        log(DEBUG, "ClientAppIo.PullClientAppInputs")
+        if self.clientapp_input is None:
+            raise ValueError(
+                "ClientAppIoInputs not set before calling `PullClientAppInputs`."
+            )
+        if request.token != self.clientapp_input.token:
+            context.abort(
+                grpc.StatusCode.INVALID_ARGUMENT,
+                "Mismatch between ClientApp and SuperNode token",
+            )
+        return PullClientAppInputsResponse(
+            message=message_to_proto(self.clientapp_input.message),
+            context=context_to_proto(self.clientapp_input.context),
+            run=run_to_proto(self.clientapp_input.run),
+        )
+
+    def PushClientAppOutputs(
+        self, request: PushClientAppOutputsRequest, context: grpc.ServicerContext
+    ) -> PushClientAppOutputsResponse:
+        """Push Message and Context."""
+        log(DEBUG, "ClientAppIo.PushClientAppOutputs")
+        if self.clientapp_output is None:
+            raise ValueError(
+                "ClientAppIoOutputs not set before calling `PushClientAppOutputs`."
+            )
+        if self.clientapp_input is None:
+            raise ValueError(
+                "ClientAppIoInputs not set before calling `PushClientAppOutputs`."
+            )
+        if request.token != self.clientapp_input.token:
+            context.abort(
+                grpc.StatusCode.INVALID_ARGUMENT,
+                "Mismatch between ClientApp and SuperNode token",
+            )
+        try:
+            # Update Message and Context
+            self.clientapp_output.message = message_from_proto(request.message)
+            self.clientapp_output.context = context_from_proto(request.context)
+            # Set status
+            code = typing.ClientAppOutputCode.SUCCESS
+            status = typing.ClientAppOutputStatus(code=code, message="Success")
+            proto_status = clientappstatus_to_proto(status=status)
+            return PushClientAppOutputsResponse(status=proto_status)
+        except Exception as e:  # pylint: disable=broad-exception-caught
+            log(ERROR, "ClientApp failed to push message to SuperNode, %s", e)
+            code = typing.ClientAppOutputCode.UNKNOWN_ERROR
+            status = typing.ClientAppOutputStatus(code=code, message="Push failed")
+            proto_status = clientappstatus_to_proto(status=status)
+            return PushClientAppOutputsResponse(status=proto_status)
+
+    def set_inputs(self, clientapp_input: ClientAppIoInputs) -> None:
+        """Set ClientApp inputs."""
+        log(DEBUG, "ClientAppIo.SetInputs")
+        self.clientapp_input = clientapp_input
+        if self.clientapp_output is not None:
+            self.clientapp_output = None
+            warnings.warn(
+                "ClientAppIoOutputs must be None at the start, setting to None",
+                stacklevel=2,
+            )
```

Comment:
We should do this check before setting the input and raise an exception if either input or output are still set. It's a clear implementation error if this occurs.